cmds/core/gosh/completer_common.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !tinygo && !plan9 && !goshsmall
+// +build !tinygo,!plan9,!goshsmall
+
+package main
+
+import (
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"unicode"
+
+	"mvdan.cc/sh/v3/syntax"
+)
+
+func lastWord(parser *syntax.Parser, line string) (int, string) {
+	var pos int
+	var endPos int
+	var word string
+	withoutSpaces := strings.TrimRightFunc(line, unicode.IsSpace)
+	// This is a giant hack, probably.
+	parser.Stmts(strings.NewReader(line), func(stmt *syntax.Stmt) bool {
+		//syntax.DebugPrint(os.Stderr, stmt)
+		if stmt.Semicolon.IsValid() {
+			// Don't autocomplete after "<stmt> &"
+			if stmt.Background || stmt.Coprocess {
+				return true
+			}
+			// We're at "<stmt>;  " with an arbitrary number of
+			// spaces after the semicolon, which we want to
+			// preserve.
+			pos = int(stmt.Semicolon.Offset()) - len(withoutSpaces) + len(line) + 1
+			endPos = int(stmt.Semicolon.Offset()) - len(withoutSpaces) + len(line) + 1
+			word = ""
+			return true
+		}
+		syntax.Walk(stmt, func(n syntax.Node) bool {
```

Comment:
I suspect walking is not what you want here, as it recurses down and you lose sight of the entire word. For example, if my last word is `"foo"bar`, you walk a DblQuoted with the Lit `foo` inside, then you walk the Lit `bar`. You would be autocompleting the last word `bar` when in reality the whole word is `foobar` when expanded.

Interactive shells like bash seem to expand the word to then autocomplete it. For example, if you have a directory called `Mustic` and you do `"Musi"<tab>`, you seem to end up with `Music/`, not `"Musi"c/` or anything else. With the sh module, you would typically expand words into fields via https://pkg.go.dev/mvdan.cc/sh/v3/expand#Fields, for example by checking that your statement is a CallExpr, expanding the last argument, and then working on that.

Just beware of the edge case of words expanding into multiple fields, e.g. what would happen if I did `Mus{x,i}<tab>` in the example above. Bash seems to give up, which is sensible.

Another edge case is expansions which need some input or state, such as environment variables or command substitutions. Bash does seem to expand `$HO<tab>` into `$HOME`, but that seems like a special case as it doesn't actually replace the input with the expanded path. `$(echo Musi)<tab>` doesn't do anything, thankfully.

Reply:
thanks. there are too many bash features in the world.

Reply:
done